src/table.rs

```diff
@@ -58,35 +63,139 @@ pub struct TableInner {
+        f.write(&data)?;
+        // TODO: pass file object directly to open and sync write
+        drop(f);
+        Self::open(path, opts)
     }
 
     /// Open an existing SST on disk
-    pub fn open(_path: &Path, _opt: Options) -> Result<TableInner> {
-        unimplemented!()
+    fn open(path: &Path, opts: Options) -> Result<TableInner> {
+        let f = fs::OpenOptions::new()
+            .read(true)
+            .write(false)
+            .create(false)
+            .open(path)?;
+        let file_name = path.file_name().unwrap().to_str().unwrap();
+        let id = parse_file_id(file_name)?;
+        let meta = f.metadata()?;
+        let table_size = meta.len();
+        let mut inner = TableInner {
+            file: MmapFile::File {
+                file: Mutex::new(f),
+                name: path.to_path_buf(),
+            },
+            table_size: table_size as usize,
+            smallest: Bytes::new(),
+            biggest: Bytes::new(),
+            id,
+            checksum: Bytes::new(),
+            estimated_size: 0,
+            index: TableIndex::default(),
+            index_start: 0,
+            index_len: 0,
+            opts,
+        };
+        inner.init_biggest_and_smallest()?;
+        // TODO: verify checksum
+        Ok(inner)
     }
 
     /// Open an existing SST from data in memory
-    pub fn open_in_memory(_data: Bytes, _id: u64, _opt: Options) -> Result<TableInner> {
-        unimplemented!()
+    fn open_in_memory(data: Bytes, id: u64, opts: Options) -> Result<TableInner> {
+        let table_size = data.len();
+        let mut inner = TableInner {
+            file: MmapFile::Memory { data },
+            opts,
+            table_size,
+            id,
+            smallest: Bytes::new(),
+            biggest: Bytes::new(),
+            checksum: Bytes::new(),
+            estimated_size: 0,
+            index: TableIndex::default(),
+            index_start: 0,
+            index_len: 0,
+        };
+        inner.init_biggest_and_smallest()?;
+        Ok(inner)
     }
 
     fn init_biggest_and_smallest(&mut self) -> Result<()> {
-        unimplemented!()
+        let ko = self.init_index()?;
+        self.smallest = Bytes::copy_from_slice(&ko.key);
+        let mut it = TableIterator::new(&self, ITERATOR_REVERSED | ITERATOR_NOCACHE);
+        it.rewind();
+        if !it.valid() {
+            return Err(Error::TableRead(format!(
+                "failed to initialize biggest for table {}",
+                self.filename()
+            )));
+        }
+        self.biggest = Bytes::copy_from_slice(it.key());
+        Ok(())
     }
 
     fn init_index(&mut self) -> Result<&BlockOffset> {
-        unimplemented!()
+        let mut read_pos = self.table_size;
+
+        // read checksum length from last 4 bytes
+        read_pos -= 4;
+        let mut buf = self.read(read_pos, 4)?;
+        let checksum_len = buf.get_u32() as usize;
+        if (checksum_len as i32) < 0 {
```

Comment:
Is it necessary?

Reply:
Seems not necessary as we're using usize now.